Draft: Updating Lotto Ecosystem (breaking changes)

Nicca42 opened the pull request:
# Changes included in this PR:

*This PR will include multiple breaking changes to the PancakeSwap Lottery contracts. These changes will greatly improve the usability and data availability within the contracts.*
---
* [x] Moving the repo over to using hardhat (from truffle) 
---
* [x] Updating the `LottoNFT.sol` from ERC721 standard to the ERC1155 standard
* [x] Allowing for batch buying of tickets 
* [x] Batch viewing for tickets 
---
* [x] Allow for a discount on batch buying
* [x] Using a three level bucket system for discounts 
---
* [x] Updating the creation of lottos
* [x] Creating a lotto now takes in the following:
* [x] `uint8 _numberOfWinningNumbers`: The number of numbers in the lotto
* [x] `uint8[] _prizeDistribution`: An array that determines the distribution 
* [x] `uint256 _prizePool`: The amount of cake available for winners
* [x] `uint256 _startingBlock`: Block time for the start of the lotto. Users cannot submit numbers before this time
* [x] `uint256 _closingBlock`: Block time stamp for end of lotto submitions. Users cannot submit numbers after this time
* [x] `uint256 _endOfLottoBlock`: Block time stamp for the end of the lotto. Users can only withdraw winnings after this time
* [x]  Returns the lotto ID. The ID will be used to determine what lotto is being referenced through the contracts
---
* [x] Updating the randomness logic to use Chain Link
---
Additional changes specified within the attached scope of work document:

[Scope of Work_ PancakeSwap Lotto.pdf](https://github.com/pancakeswap/lottery-contract/files/6036545/Scope.of.Work_.PancakeSwap.Lotto.pdf)

Nicca42 committed:
Lotto: Styling & org

Nicca42 committed:
Lotto: Styling & org

Nicca42 committed:
Added hardhat

Nicca42 committed:
Added hardhat

Nicca42 committed:
Merge branch 'vonnie610/adding-hardhat'

Nicca42 committed:
NFT started

Nicca42 committed:
NFT started

Nicca42 committed:
Creating new lotto

Nicca42 committed:
Creating new lotto

Nicca42 committed:
Fixed gas for storage on Lotto

Nicca42 committed:
Fixed gas for storage on Lotto

Nicca42 committed:
Drawing numbers temp mock for chainlink VRF

Nicca42 committed:
Drawing numbers temp mock for chainlink VRF

Nicca42 committed:
Moved storage to NFT

Nicca42 committed:
Moved storage to NFT

Nicca42 committed:
Added view functionality to NFT

Nicca42 committed:
Added view functionality to NFT

Nicca42 committed:
Gas optimisation

Nicca42 committed:
Gas optimisation

Nicca42 committed:
Merge branch 'vonnie610/lotto-nft-to-erc1155'

Nicca42 committed:
discount

Nicca42 committed:
discount

ChefNyan commented on contracts/Lotto.sol:
```diff
@@ -0,0 +1,287 @@
+        maxValidRange_ = _maxValidNumberRange;
+    }
+
+    function init(address _lotteryNFT) public onlyOwner() {
```
What's the benefit of initializing separately? To use a proxy, we should use initialize and remove all from constructor. https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable

Nicca42 replied:
When I add the Proxy contract I will address this.

ChefNyan commented on contracts/Lotto.sol:
```diff
@@ -0,0 +1,287 @@
+        onlyOwner()
+        returns(uint256 lottoID)
+    {
+        lotteryIDCounter_.increment();
```
It needs a check to make sure `_prizeDistribution` cannot be higher than 100%

Nicca42 replied:
Added check

ChefNyan commented on contracts/LottoNFT.sol:
```diff
@@ -0,0 +1,192 @@
+    // EVENTS
+    //-------------------------------------------------------------------------
+
+    event infoBatchMint(
```
`InfoBatchMint`

Nicca42 replied:
Updated all events to have a capital start.

ChefNyan commented on contracts/LottoNFT.sol:
```diff
@@ -0,0 +1,192 @@
+        uint256[] memory _tokenIDs, 
+        uint256[] memory _amounts
+    ) 
+        public 
```
why is the scope `public` if it is `batchMintInternal`

Nicca42 replied:
Made it public to test. This was when I was trying to make a more efficient batch mint function. It has been removed.

ChefNyan commented on contracts/LottoNFT.sol:
```diff
@@ -0,0 +1,192 @@
+        uint32 _numberOfTickets,
+        uint32[] memory _numbers
+    )
+        public
```
putting it external should reduce the cost (and put `_numbers` as calldata)

Nicca42 replied:
Updated all public functions to external. Updated to `calldata`

ChefNyan commented on contracts/Lotto.sol:
```diff
@@ -0,0 +1,287 @@
+    function drawWinningNumbers(uint256 _lottoID) public onlyOwner() {
+        // Creating space for winning numbers
+        uint8[] memory winningNumbers;
+        // TODO will call ChainLink VRF
```
?

Nicca42 replied:
Waiting on the green from Chef Hops for ChainLink integration BSC. 

In the meantime going to make a mock contract + interface

ChefNyan commented on contracts/Lotto.sol:
```diff
@@ -0,0 +1,287 @@
+     *          will be pushed back until winning numbers are added. 
+     */
+    function createNewLotto(
+        uint8[] calldata _prizeDistribution,
```
calldata is for external functions

Nicca42 replied:
Call now external

ChefNyan commented on contracts/Lotto.sol:
```diff
@@ -0,0 +1,287 @@
+    /**
+      * @return Gets the current time
+      */
+    function getTime() public view returns(uint256) {
```
what's the purpose of this function? It can be obtained from the blockchain itself

Nicca42 replied:
Had it there out of habit, other UI devs I've worked with like having a function for it. Will remove.

ChefNyan commented on contracts/Lotto.sol:
```diff
@@ -0,0 +1,287 @@
+    // Storing of the NFT
+    ILottoNFT internal nft_;
+
+    // Lottery size
+    uint8 internal sizeOfLottery_;
+    // Max range for numbers (starting at 0)
+    uint8 internal maxValidRange_;
```
it can be public

Nicca42 replied:
I'm assuming you want them both to be public. Will make it so, let me know if you want anything different.

ChefNyan commented on cache/solidity-files-cache.json:
```diff
@@ -0,0 +1,1341 @@
+{
```
This one can be added to .gitignore

Nicca42 replied:
`cache` added to `.gitignore`

ChefNyan commented on contracts/Lotto.sol:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    function batchBuyLottoTicket(
+        uint256 _lotteryID,
```
you can buy for old lotteries (winning by knowing the numbers, then)

Nicca42 replied:
Check added

Nicca42 committed:
Resolving comments 1

Nicca42 committed:
Resolving comments 1

Nicca42 committed:
WIP Updating tests with new time

Nicca42 committed:
WIP Updating tests with new time

Nicca42 committed:
Added test coverage

Nicca42 committed:
Added test coverage

Nicca42 committed:
Fixed time issue in tests

Nicca42 committed:
Fixed time issue in tests

Nicca42 committed:
ChainLink - version issue

Chainlink VRFConsumerBase is Solidity 0.6.0, so Lottery contract cannot inherit from it as its 0.7.3 and this is an incompatible version jump. Would require updating Chainlink contracts version.

Nicca42 committed:
ChainLink - version issue

Chainlink VRFConsumerBase is Solidity 0.6.0, so Lottery contract cannot inherit from it as its 0.7.3 and this is an incompatible version jump. Would require updating Chainlink contracts version.

Nicca42 committed:
Merge branch 'vonnie610/before-chainlink-version-issue' into vonnie610/claim-logic

Nicca42 committed:
Cleaned up creating lotto function

Nicca42 committed:
Cleaned up creating lotto function

Nicca42 committed:
WIP Testing claim logic

Nicca42 committed:
WIP Testing claim logic

Nicca42 committed:
WIP Claiming Logic tests

Nicca42 committed:
WIP Claiming Logic tests

Nicca42 committed:
Added batch claim + test

Nicca42 committed:
Added batch claim + test

Nicca42 committed:
Merge branch 'vonnie610/claim-logic'

Nicca42 committed:
Merge branch 'master' into vonnie610/chainlink

Nicca42 committed:
Stash

Nicca42 committed:
Stash

Nicca42 committed:
Stash 2

Nicca42 committed:
Stash 2

Nicca42 committed:
Added random generator

Nicca42 committed:
Added random generator

Nicca42 committed:
Stash for testing

Nicca42 committed:
Stash for testing

Nicca42 committed:
Stash - other changes

Nicca42 committed:
Stash - other changes

Nicca42 committed:
Updates from list 1

Nicca42 committed:
Updates from list 1

Nicca42 committed:
Converted numbers from uint32 to uint16

Nicca42 committed:
Converted numbers from uint32 to uint16

Nicca42 committed:
Moved ticket claim check to NFT

Nicca42 committed:
Moved ticket claim check to NFT

Nicca42 committed:
Made external

Nicca42 committed:
Made external

Nicca42 committed:
Added discount + upgrade functionality

Nicca42 committed:
Added discount + upgrade functionality

Nicca42 committed:
All lotto functionality present. WIP tests

Nicca42 committed:
All lotto functionality present. WIP tests

Nicca42 committed:
All tests working

Nicca42 committed:
All tests working

Nicca42 committed:
Final changes + tests + docs

Nicca42 committed:
Final changes + tests + docs

Nicca42 committed:
Merge branch 'vonnie610/chainlink'

ChefNyan commented on .gitignore:
```diff
@@ -1,6 +1,7 @@
-node_modules/*
-contracts/flattened/*.sol
\ No newline at end of file
+.idea/
```
please add cache too

Nicca42 replied:
The cache is already in the gitignore. Not sure why it is not being ignored, might be a hardhat thing.

ChefNyan commented on contracts/LotteryNFT.sol:
```diff
@@ -1,59 +1,211 @@
+    }
+    // Token ID => Token information 
+    mapping(uint256 => TicketInfo) internal ticketInfo_;
+    mapping(address => uint256[]) internal userTickets_;
```
can we add a double mapping per lotteryId, too?

Nicca42 replied:
The max amount of tokens we can mint right now is 50. If I add another mapping that number will be reduced greatly.

ChefNyan commented on contracts/LotteryNFT.sol:
```diff
@@ -1,59 +1,211 @@
+    event InfoBatchMint(
+        address indexed receiving, 
+        uint256 amountOfTokens, 
+        uint256[] tokenIds
```
It should also incorporate `lotteryId` (it will help for a subgraph)

Nicca42 replied:
added

ChefNyan commented on contracts/LotteryNFT.sol:
```diff
@@ -1,59 +1,211 @@
-        claimInfo[tokenId] = true;
+
+    function getUserTickets(address _user) external view returns(uint256[] memory) {
+        return userTickets_[_user];
```
it can return "out of gas" if too large, if the `lotteryId` is added + pagination, it will prevent this

Nicca42 replied:
We don't have the ability here to filter by lottery ID, as mentioned in the other unresolved review, the gas cost for adding the lottery ID into the structs would be significant.

ChefNyan commented on contracts/LotteryNFT.sol:
```diff
@@ -1,59 +1,211 @@
+     */
+    function batchMint(
+        address _to,
+        uint256 _lottoID,
```
is this one used?

Nicca42 replied:
Yes

ChefNyan commented on contracts/RandomNumberGenerator.sol:
```diff
@@ -0,0 +1,69 @@
+        uint256 lotteryId,
+        uint256 userProvidedSeed
+    ) 
+        public 
```
Anyone can call it so it means EOA could empty the LINK token balance

ChefNyan commented on hardhat.config.js:
```diff
@@ -0,0 +1,37 @@
+  gasReporter: {
+    enabled: true,
+    currency: 'CHF',
+    gasPrice: 21
```
it should be 10 for BSC

ChefNyan commented on hardhat.config.js:
```diff
@@ -0,0 +1,37 @@
+module.exports = {
+  gasReporter: {
+    enabled: true,
+    currency: 'CHF',
```
'USD'

ChefNyan commented on hardhat.config.js:
```diff
@@ -0,0 +1,37 @@
+        interval: 5000
+      },
+      blockGasLimit: 13000000,
+      gasPrice: 20
```
10

ChefNyan commented on contracts/Lottery.sol:
```diff
@@ -1,393 +1,648 @@
+    Counters.Counter private lotteryIDCounter_;
+    // Safe math
     using SafeMath for uint256;
+    using SafeMath for uint16;
```
SafeMath is only for uint256

ChefNyan commented on contracts/Lottery.sol:
```diff
@@ -1,393 +1,648 @@
+
+    event NewLotteryCreated(
+        uint256 indexed lotteryId,
+        Status lotteryStatus,
```
I think it can be removed since the status is always the same for all of these events

ChefNyan commented on contracts/Lottery.sol:
Can we add an event, please?

Nicca42 replied:
added event

ChefNyan commented on contracts/Lottery.sol:
Can we also add an event here, pls?

Nicca42 replied:
added event

ChefNyan commented on contracts/Lottery.sol:
it is better if `_amount` is set as a parameter

Nicca42 replied:
Updated

ChefNyan commented on contracts/Lottery.sol:
where is this one initialized?

Nicca42 replied:
The winning numbers? They are initialised and set in the internal split function.

ChefNyan commented on contracts/Lottery.sol:
it doesn't check the `lotteryId` of the `tokenId`. so it may be possible to use old tickets for future lotteries, I think

Nicca42 replied:
Nope. If the ticket was not bought for the specific lottery it will revert above this:
```
require(
            nft_.claimTicket(_tokenId, _lotteryId),
            "Numbers for ticket invalid"
        );
```

The `NFT.claimTicket` checks if the ticket is for the lottery. Will not get to the number of matching line.

Nicca42 committed:
Added lottery ID check and random number modfier

Nicca42 committed:
Added lottery ID check and random number modfier

Nicca42 committed:
Comment changes

Nicca42 committed:
Comment changes

Nicca42 committed:
Lottery ID on NFT contract

Nicca42 committed:
Lottery ID on NFT contract

Nicca42 committed:
Ticket ID returns test

Nicca42 committed:
Ticket ID returns test

ChefNyan commented on contracts/LotteryNFT.sol:
```diff
@@ -1,59 +1,236 @@
+            ticketInfo_[_ticketID].lotteryId == _lotteryId,
+            "Ticket not for this lottery"
+        );
+        uint256 maxRange = ILottery(lotteryContract_).getMaxRange();
```
it may cause problem if the lotteryId's range changes. It should be read from a function that includes the `lotteryId`

Nicca42 committed:
Audit changes - tests not updated

Nicca42 committed:
Audit changes - tests not updated

Nicca42 committed:
Updated changes

Nicca42 committed:
Updated changes

Nicca42 committed:
Added custom SafeMath for uint16 & uint8

Nicca42 committed:
Added custom SafeMath for uint16 & uint8

Nicca42 committed:
Removed redundant check

Nicca42 committed:
Removed redundant check

Nicca42 committed:
Added pagination

Nicca42 committed:
Added pagination

Nicca42 committed:
Removed coverage

Nicca42 committed:
Removed coverage

Nicca42 committed:
Added total supply and updated tests

Nicca42 committed:
Added total supply and updated tests

Nicca42 committed:
Removed unused variable

Nicca42 committed:
Removed unused variable

Nicca42 committed:
Rebase and removed cache file and updated gitignore

Nicca42 committed:
resolved merge issue
